lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py

```diff
@@ -0,0 +1,416 @@
+    "default": {},
+    "examples": [
+        {
+            "application-data": {
+                "url": "https://grafana.example.com/"
+            }
+        }
+    ],
+    "required": [
+        "application-data"
+    ],
+    "properties": {
+        "application-data": {
+            "$id": "#/properties/application-data",
+            "title": "Application Databag",
+            "type": "object",
+            "additionalProperties": True,
+            "required": [
+                "url"
+            ],
+            "url": {
+                "$id": "#/properties/application-data/url",
+                "type": "string"
+            }
+        }
+    },
+    "additionalProperties": True
+}
+
+logger = logging.getLogger(__name__)
+
+
+class UrlAvailableEvent(EventBase):
+    """Charm event triggered when provider charm extracts the url from relation data"""
+
+    def __init__(self, handle, url: str, relation_id: int):
+        super().__init__(handle)
+        self.url = url
+        self.relation_id = relation_id
+
+    def snapshot(self) -> dict:
+        """Returns snapshot."""
+        return {
+            "url": self.url,
+            "relation_id": self.relation_id,
+        }
+
+    def restore(self, snapshot: dict):
+        """Restores snapshot."""
+        self.url = snapshot["url"]
+        self.relation_id = snapshot["relation_id"]
+
+
+class AuthProviderCharmEvents(CharmEvents):
+    """List of events that the auth provider charm can leverage."""
+
+    url_available = EventSource(UrlAvailableEvent)
+
+class GrafanaAuthProvides(Object):
+    """Grafana authentication configuration provider class to be initialized by grafana-auth providers"""
+
+    on = AuthProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str, auth_type: str, **kwargs):
+        super().__init__(charm, relationship_name)
+        self._charm = charm
+        self._relationship_name = relationship_name
+        self._auth_conf = self._build_conf_dict(auth_type, **kwargs)
+        self.framework.observe(
+            charm.on[relationship_name].relation_joined, self._set_auth_config_in_relation_data
+        )
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._on_auth_relation_changed
+        )
+
+    def _build_conf_dict(self, auth_type, **kwargs) -> dict:
+        """Builds a dictionary for authentication configuration that matches the json schema of the provider.
+        Args:
+            auth_type (str): authentication type to be set in the configuration dict.
+            kwargs: key and value pairs provided to configure authentication mode.
+        Returns:
+            dict: Authentication configuration as a dictionary.
+        """
+        return {auth_type: { key:value for key, value in kwargs.items()}}
+
+    def _set_auth_config_in_relation_data(self, event: RelationJoinedEvent) -> None:
```

Comment:
Would this eventually be also called on leader-elected?
If the relation event takes place while the leader is unreachable but before a new one was elected, this reldata would never be set. ([example](https://github.com/canonical/prometheus-k8s-operator/blob/e75eb0b1a7669c945dabb6b4ba8cfe583697533b/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1517))

Reply:
```suggestion
    def _set_auth_config_in_relation_data(self, event: RelationJoinedEvent) -> None:
        if not self._charm.unit.is_leader():
            event.defer()
            return
```
Would adding `event.defer()` be a good solution instead of calling the function on both events?

Reply:
I realised that it doesn't make sense to use `event.defer()` here, so the handler will run on `leader elected` as you suggested. Fixed :)